### PR TITLE
Align docker and vagrant base images

### PR DESCRIPTION
**Pull request by zvecr**
## Description

Just noticed docker is using Debian, while vagrant is using Ubuntu. Its easy enough to change either way, but to quote a reply from my question on discord, "Debian is the de facto standard for containers"

While I've only just updated to the "bento/ubuntu-16.04" box, the main aim was to use a box that was updated more frequently and both seem to be updated monthly. Given it was previously broken, now seems a good time to make this change.

## Types of changes

- [ ] Core
- [ ] Bugfix
- [ ] New Feature
- [x] Enhancement/Optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/Layout/Userspace (addition or update)
- [ ] Documentation

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -6,7 +6,7 @@ Vagrant.configure(2) do |config|
   config.vm.define "qmk_firmware"
 
   # VMware/Virtualbox ( and also Hyperv/Parallels) 64 bit
-  config.vm.box = "bento/ubuntu-16.04"
+  config.vm.box = "generic/debian9"
```
Comment (vomindoraan) [inline Vagrantfile]: Is there a `generic/debian`, i.e. latest, with no version number? That way the Dockerfile (which uses `FROM debian`) and the Vagrantfile will keep themselves up to date, and we won't need to update this once Debian 10 is out (which will be in a couple of months).

Comment (zvecr) [reply Vagrantfile]: The official debian boxes use the code name, [like debian/stretch64](https://app.vagrantup.com/debian/boxes/stretch64) and they only build for virtualbox, not VMware which has been in the Vagrantfile for some time. We could consider locking the Docker image to `debian:9` if that helps keep things consistent?

Comment (vomindoraan) [reply Vagrantfile]: Yes, that would help.

Comment (zvecr) [reply Vagrantfile]: Done, I've also noted the image size is rather large but i will tackle that in a separate PR.

**Comment by zvecr** [comment]
Anything i can provide (logs,etc) to progress this?

**Comment by drashna** [comment]
@jackhumbert @skullydazed ?

**Comment by zvecr** [comment]
As we have an issue raised against the fact that ubuntu is using an old version of git, #5083, is there any chance this could be reviewed/merged?

**Comment by drashna** [comment]
Well, we're using an old version of ubuntu there, as well (16), and 18.04 is a no-go right now (ARM toolchain is broken). 

Though, this is much more up to skully/jack/etc, as they're more familiar with this stuff.

**Comment by yanfali** [comment]
Upvote for debian, it's more stable and better maintained than Ubuntu who do arbitrary time based snapshotting.

**Comment by zvecr** [comment]
Just noticed the vagrantfile docker provider also needs updating, i will fix and test shortly.

**Comment by skullydazed** [comment]
+1 to Debian

**Comment by zvecr** [comment]
I could not find a nice docker image that can be used with vagrant (updated frequently, many downloads, reliable publisher, etc). The requirement from vagrant is that ssh is configured and running, which the base `debian:9` image is not. There is support for building a docker image, which is built on demand, however i would consider this out of scope for now.

I have instead updated the vagrant docker image version to one that at least matches.

**Comment by yanfali** [comment]
@zvecr can you pin the docker image using a SHA please? I've seen upstream providers change the image and break it by making _minor_ changes.

**Comment by yanfali** [comment]
https://medium.com/@tariq.m.islam/container-deployments-a-lesson-in-deterministic-ops-a4a467b14a03 for details on how to do this

**Comment by zvecr** [comment]
@yanfali my experience with docker is that pinning varies on the target consumer.
For most things i have found it best to pin to major or minor versions, you still get bug/security fixes but nothing major _should_ get upgraded under your feet. For images where they contain non OS dependencies, like node or nginx, and you have developed against a version, i found it safer to pin to minor versions.

Only a few situations were considered for pinning to explicit versions, but these were ruled out due to the maintenance involved, even when the consumer had adequate CI coverage. #4862 could help with this, as would smoke tests within the docker hub automated build.

Happy to go with whatever the group consensus is.

**Comment by yanfali** [comment]
I get what you're saying, but in my experience the tags have no real value, and the image can be replaced at any time, which leads to awkward situations where someone updates an image and re-uses a tag and doesn't do enough testing. Are we going to use these on our build infrastructure? or is this purely for devs to build with?

**Comment by zvecr** [comment]
This change in isolation, it purely dev focused. #4862 is where it starts to transition, however there is no reason why that PR cant change. Locally i have a few different strategies, including multistage builds.

**Comment by yanfali** [comment]
Sounds good. For Travis, I think we should lock down the image, but for dev builds I think you are right it will be ok.

**Comment by zvecr** [comment]
@yanfali Thank for the input. Happy to resume conversation within #4862, RE local reproducible builds.

**Comment by yanfali** [comment]
@drashna this looks fairly safe to merge

**Pull request merged**